common4j/src/main/com/microsoft/identity/common/java/cache/SharedPreferencesAccountCredentialCache.java

```diff
@@ -100,113 +107,150 @@ public SharedPreferencesAccountCredentialCache(
+            } finally {
+                mLoaded = true;
+                mCacheLock.notifyAll();
+            }
+        }
+    }
+
+    private void waitForInitialLoad() {
+        while (!mLoaded) {
+            try {
+                mCacheLock.wait();
+            } catch (InterruptedException e) {}
+        }
     }
 
     @Override
-    public synchronized void saveAccount(@NonNull final AccountRecord accountToSave) {
-        Logger.verbose(TAG, "Saving Account...");
-        Logger.verbose(TAG, "Account type: [" + accountToSave.getClass().getSimpleName() + "]");
+    public void saveAccount(@NonNull final AccountRecord accountInput) {
+        final String methodTag = TAG + ":saveAccount";
+
+        AccountRecord accountToSave = null;
+        try {
+            accountToSave = (AccountRecord) accountInput.clone();
+        } catch (CloneNotSupportedException e) {
+            Logger.error(methodTag, "Failed to clone AccountRecord", e);
+            return;
+        }
+
+        Logger.verbose(methodTag, "Saving Account...");
+        Logger.verbose(methodTag, "Account type: [" + accountToSave.getClass().getSimpleName() + "]");
         final String cacheKey = mCacheValueDelegate.generateCacheKey(accountToSave);
-        Logger.verbosePII(TAG, "Generated cache key: [" + cacheKey + "]");
+        Logger.verbosePII(methodTag, "Generated cache key: [" + cacheKey + "]");
 
-        // Perform any necessary field merging on the Account to save...
-        final AccountRecord existingAccount = getAccount(cacheKey);
+        synchronized (mCacheLock) {
+            waitForInitialLoad();
 
-        if (null != existingAccount) {
-            accountToSave.mergeAdditionalFields(existingAccount);
-        }
+            // Perform any necessary field merging on the Account to save...
+            final AccountRecord existingAccount = getAccount(cacheKey);
 
-        final String cacheValue = mCacheValueDelegate.generateCacheValue(accountToSave);
-        mSharedPreferencesFileManager.put(cacheKey, cacheValue);
+            if (null != existingAccount) {
+                accountToSave.mergeAdditionalFields(existingAccount);
+            }
+
+            final String cacheValue = mCacheValueDelegate.generateCacheValue(accountToSave);
+            mSharedPreferencesFileManager.put(cacheKey, cacheValue);
+            mCachedAccountRecordsWithKeys.put(cacheKey, accountToSave);
+        }
     }
 
     @Override
-    public synchronized void saveCredential(@NonNull Credential credentialToSave) {
-        Logger.verbose(TAG, "Saving credential...");
+    public void saveCredential(@NonNull Credential credentialInput) {
+        final String methodTag = TAG + ":saveCredential";
+
+        Credential credentialToSave = null;
+        try {
+            credentialToSave = (Credential) credentialInput.clone();
+        } catch (CloneNotSupportedException e) {
+            Logger.error(methodTag, "Failed to clone Credential", e);
```

Comment:
So we are not going to save the credential in this case? Should we just throw an Exception rather than returning silently out of this method?

Reply:
See above